src/lib/fetch-document-list.js

```diff
@@ -0,0 +1,47 @@
+const { info } = require('../lib/log');
+const fs = require('../lib/sync-fs');
+const DOC_TYPES = ['district_hospital', 'health_center', 'clinic', 'person', 'user', 'user-settings', 'contact'];
+
+const jsonDocPath = (directoryPath, docID) => `${directoryPath}/${docID}.doc.json`;
+const fetchDocumentListFromDB = async (db, ids) => {
+  info('Downloading doc(s)...');
+  const documentDocs = await db.allDocs({
+    keys: ids,
+    include_docs: true,
+  });
+
+  const missingDocumentErrors = documentDocs.rows.filter(row => !row.doc).map(row => `Document with id '${row.key}' could not be found.`);
+  if (missingDocumentErrors && missingDocumentErrors.length) {
+    throw Error(missingDocumentErrors);
+  }
+
+  const documentTypeErrors = documentDocs.rows.filter(row => !DOC_TYPES.includes(row.doc.type)).map(row => ` Document with id ${row.key} of type ${row.doc.type} cannot be edited`);
+  if (documentTypeErrors && documentTypeErrors.length) {
+    throw Error(documentTypeErrors);
```

Comment:
I think a way to improve performance is looping through the docs to find errors once, because `filter` will run through all the records anyways, so instead of looping 4 times (filter, map and then filter map again) we can use `forEach`. (Thinking if we need to process many docs)

```js
const errors =  { invalidDocType: [], docNotFound: [] };

documentDocs.rows.forEach(row => {
   if (!row.doc) {
      errors.docNotFound.push(`Document with id '${row.key}' could not be found.`);
      return;
   }

   if (!DOC_TYPES.includes(row.doc.type)) {
     errors.invalidDocType.push(`Document with id ${row.key} of type ${row.doc.type} cannot be edited`);
   }
});

// then since we initialised the errors arrays, we just check of length
if (errors.docNotFound.length || errors.invalidDocType.length) {
   throw Error([...errors.docNotFound, ...errors.invalidDocType]); // errors are printed in order, all the not found and then the invalid type. Check if we can do [...var] in node versions
}
```

Reply:
And once the user fixes all error then we preform the success path, as you have it in L23 :)

Reply:
Good suggestion!